hip-clang.docker

```diff
@@ -74,6 +74,8 @@ RUN cget -p $PREFIX init --cxx /opt/rocm/llvm/bin/clang++ --std=c++14 -DAMDGPU_T
 
 # Install dependencies
 RUN cget -p $PREFIX install pfultz2/rocm-recipes
+# Install a newer version of cmake for libMLIRMIOpen
```

Comment:
Each change of the dockerfile leads to docker cache invalidation, so the new docker image will be built on CI node. This would lead to delays in other CI jobs running in parallel. In order to minimize impact, we shall update basename of the docker image whenever we change the docker file.

Please go here:https://github.com/ROCmSoftwarePlatform/MIOpen/blob/896e823c0fb90db4f2ecd9c4fec47b7c4c6eb050/Jenkinsfile#L237
and change "miopen-hip-clang" to "miopen-v1".

Reply:
I think there is a misunderstanding how docker works, changing the image name will not make any difference.

Each of the commands in a dockerfile represent a docker layer(as it lays each of the changes on top of each other). Locally, docker caches all these layers. So if you change a command later in the dockerfile it will reuse the previous docker layers by default. 

Each docker layer is essentially a docker image named after a sha256 hash that uniquely identifies it. You can run these previous layers if you know the hash of that layer.

Now, when we name an image it essentially tags it by associating that name with the hash of the last layer in the dockerfile. If you later, tag it to a different hash, it will simply reassociate the name to a different hash. It will not delete the unused docker layers, so if a later PR switches the tag back to the previous value, docker will simply use the cached image instead of rebuilding.

The [`docker prune`](https://docs.docker.com/config/pruning/) command is needed to remove these unused docker layers. From the documentation:

> Docker takes a conservative approach to cleaning up unused objects (often referred to as “garbage collection”), such as images, containers, volumes, and networks: these objects are generally not removed unless you explicitly ask Docker to do so. 

You can also prove this to yourself by trying to retag an image locally. First, build the dockerfile with a tag, change the dockerfile and rebuild with the same tag, and then finally revert the change to the dockerfile and retag. Docker will reuse the previous layers.

Reply:
> Then for each code change PR, it always pulls in the latest tag from dockerhub instead of building it every time in each PR.

We would like to have reproducibility and traceability in our CI, which that prevents.

Reply:
> As long as docker hub keep the image for every commit, you still have reproducibility and traceability by inspecting the sha256. When latest tag get re-written, the image tagged with commit won't.

But if its pulling from the latest tags this means that building older jobs could be very different.

Also, the docker image is opaque and there is no visible steps that was used to build the docker image(even storing the dockefile in the repo doesn't mean its the same as the docker image being pulled).

Ideally, to improve build time is to use a centralized docker layer cache.

Reply:
> building older jobs could be very different.

That's true, unless manually update the tag to the right one.

> storing the dockefile in the repo doesn't mean its the same as the docker image being pulled

Since PR push policy enforce dockerfile change to be standalone (from any development changes), we have this guarantee.

Reply:
@pfultz2 @jerryyin I do not care how docker works in my local environment. What I see with my own eyes is that since image names harmonized there are much less docker rebuilds on our CI. Let's follow the currently adopted convention for naming images in this repo. There is no harm at least.

Reply:
> I do not care how docker works in my local environment.

This is way docker works period. Furthermore, the jenkins pipeline doesn't do anything special to conclude that a local environment would be different.

> What I see with my own eyes is that since image names harmonized there are much less docker rebuilds on our CI.

There is less rebuilds because more of the slave nodes have cached the docker images. 

> There is no harm at least.

Yes, but its important to understand how docker works so issues can be addressed without cargo cult solutions like renaming docker images.

Reply:
> so issues can be addressed without cargo cult solutions like renaming docker images.

That's true. But the question is: Why this renaming of docker images was used in the old Jenkinsfile? Do you know what was the original intent?

Reply:
@pfultz2 This is where all this stuff is originated from: https://github.com/AMDComputeLibraries/MLOpen/pull/2147/files?diff=split&w=1#r364339367

Reply:
> Why this renaming of docker images was used in the old Jenkinsfile? Do you know what was the original intent?

I am not sure exactly why it was done. I guess Daniel felt the need to name the images different, perhaps to make it easier to access the images on the nodes directly(sometime we would debug issues directly on the jenkins node). However, he never mentioned cache "invalidation" as a reason. 

Nor has cache "invalidation" ever been an issue in the past, especially when we had more limited resources in our CI, and more frequent changes to the dockerfile.

Reply:
@pfultz2 I simply thought that if you build image A using the same image name as the existing image B, then later you will need to rebuild image B from scratch. But it seems that things are different.

I will do some tests and then get rid of all cargo cult code at once, but not right away. Just do not have enough time.

Reply:
In fact, I will gladly get rid of these bells and whistles around image names.